### ISSUE TITLE
[1.13.1] Rise in "device or resource busy" errors when stopping/removing containers

### ISSUE THREAD

**Issue by BenHall** [moby/moby#31195]
I'm seeing a rise in the number of "device or resource busy" errors we're getting. It takes at least two attempts to remove certain containers. We're using overlay.

```
$ web_1
Error response from daemon: Unable to remove filesystem for af88d352ccc43d6ccb64ebef61d66d336c9a8fb11c194f21ca96ae719a402a38: remove /home/docker/data/containers/af88d352ccc43d6ccb64ebef61d66d336c9a8fb11c194f21ca96ae719a402a38/shm: device or resource busy
$ docker rm -f web_1
Error response from daemon: No such container: web_1
```

Docker Compose can get very confused at this point. 
```
$ docker rm -f web_1
web_1
$ docker rm -f web_1
Error response from daemon: No such container: web_1
$ docker-compose up -d
Creating web_1

ERROR: for web  Cannot start service web: service endpoint with name web_1 already exists
ERROR: Encountered errors while bringing up the project.
$ docker ps -a | grep web_1
fae79acf199f       web                                     "npm start"              About a minute ago   Created                                             _web_1
```

This is happening on multiple servers.  Journal is full of:
```
Container cc2b685adefd failed to exit within 10 seconds of kill - trying direct SIGKILL
```

```
$ docker-compose version
docker-compose version 1.9.0, build 2585387
docker-py version: 1.10.6
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1t  3 May 2016
```

```
$ docker version
Client:
 Version:      1.13.1
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 06:57:21 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.1
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 06:57:21 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

```
$ uname -a
Linux host07 4.9.9-040909-generic #201702090333 SMP Thu Feb 9 08:35:27 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by BenHall** [comment]
Docker stopped responding to launching new containers too. Had to reboot.

**Comment by BenHall** [comment]
More logs
```
Mar 06 14:20:23 www2 dockerd[16648]: time="2017-03-06T14:20:23.061019177+01:00" level=warning msg="container kill failed because of 'container not found' or 'no such process': Cannot kill container 2aeea274c55f702cf93d866b1b3f3c18418b6c8b51afba73016372028a540f32: rpc error: code = 2 desc = containerd: container not found"
Mar 06 14:20:33 www2 dockerd[16648]: time="2017-03-06T14:20:33.062270830+01:00" level=info msg="Container 2aeea274c55f failed to exit within 10 seconds of kill - trying direct SIGKILL"
Mar 06 14:20:33 www2 dockerd[16648]: time="2017-03-06T14:20:33.063933591+01:00" level=warning msg="container kill failed because of 'container not found' or 'no such process': Cannot kill container 2aeea274c55f702cf93d866b1b3f3c18418b6c8b51afba73016372028a540f32: rpc error: code = 2 desc = containerd: container not found"
Mar 06 14:20:36 www2 dockerd[16648]: time="2017-03-06T14:20:36.064677255+01:00" level=info msg="Container 2aeea274c55f702cf93d866b1b3f3c18418b6c8b51afba73016372028a540f32 failed to exit within 3 seconds of signal 15 - using the force"
Mar 06 14:20:36 www2 dockerd[16648]: time="2017-03-06T14:20:36.066709265+01:00" level=warning msg="container kill failed because of 'container not found' or 'no such process': Cannot kill container 2aeea274c55f702cf93d866b1b3f3c18418b6c8b51afba73016372028a540f32: rpc error: code = 2 desc = containerd: container not found"
Mar 06 14:20:46 www2 dockerd[16648]: time="2017-03-06T14:20:46.067295223+01:00" level=info msg="Container 2aeea274c55f failed to exit within 10 seconds of kill - trying direct SIGKILL"
Mar 06 14:20:46 www2 dockerd[16648]: time="2017-03-06T14:20:46.099430338+01:00" level=error msg="Handler for DELETE /v1.26/containers/2aeea274c55f returned error: Unable to remove filesystem for 2aeea274c55f702cf93d866b1b3f3c18418b6c8b51afba73016372028a540f32: remove /home/docker/data/containers/2aeea274c55f702cf93d866b1b3f3c18418b6c8b51afba73016372028a540f32/shm: device or resource busy"
```

**Comment by BenHall** [comment]
When doing a docker-compose up -d

`ERROR: for web  You cannot remove a running container 97c0acf1ad2bca329590963e0429646da59d4c5ca337ea298e289f4b43c21735. Stop the container before attempting removal or use -f`


```
Mar 06 14:23:37 www2 dockerd[16648]: time="2017-03-06T14:23:37.236918777+01:00" level=warning msg="container kill failed because of 'container not found' or 'no such process': Cannot kill container 97c0acf1ad2bca329590963e0429646da59d4c5ca337ea298e289f4b43c21735: rpc error: code = 2 desc = containerd: container not found"
Mar 06 14:23:47 www2 dockerd[16648]: time="2017-03-06T14:23:47.237911506+01:00" level=info msg="Container 97c0acf1ad2b failed to exit within 10 seconds of kill - trying direct SIGKILL"
Mar 06 14:23:48 www2 dockerd[16648]: time="2017-03-06T14:23:48.064676348+01:00" level=error msg="Handler for DELETE /v1.21/containers/97c0acf1ad2bca329590963e0429646da59d4c5ca337ea298e289f4b43c21735 returned error: You cannot remove a running container 97c0acf1ad2bca329590963e0429646da59d4c5ca337ea298e289f4b43c21735. Stop the container before attempting removal or use -f"
Mar 06 14:24:46 www2 dockerd[16648]: time="2017-03-06T14:24:46.475650010+01:00" level=info msg="Container 68c5b477ff5b132b0942fa7985fac0102ab155bac31adf7d423279ae4a7f8c70 failed to exit within 10 seconds of signal 15 - using the force"
Mar 06 14:24:46 www2 dockerd[16648]: time="2017-03-06T14:24:46.479433133+01:00" level=warning msg="container kill failed because of 'container not found' or 'no such process': Cannot kill container 68c5b477ff5b132b0942fa7985fac0102ab155bac31adf7d423279ae4a7f8c70: rpc error: code = 2 desc = containerd: container not found"
Mar 06 14:24:56 www2 dockerd[16648]: time="2017-03-06T14:24:56.480976678+01:00" level=info msg="Container 68c5b477ff5b failed to exit within 10 seconds of kill - trying direct SIGKILL"
Mar 06 14:24:56 www2 dockerd[16648]: time="2017-03-06T14:24:56.503713487+01:00" level=error msg="Handler for DELETE /v1.21/containers/68c5b477ff5b132b0942fa7985fac0102ab155bac31adf7d423279ae4a7f8c70 returned error: You cannot remove a running container 68c5b477ff5b132b0942fa7985fac0102ab155bac31adf7d423279ae4a7f8c70. Stop the container before attempting removal or use -f"
```

This is on a different / new VM.

```
$ uname -a
Linux www2 4.9.11-040911-generic #201702181031 SMP Sat Feb 18 15:33:41 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by BenHall** [comment]
Can Docker rename containers? It looks like a process is still running, but with a different name.

```
97c0acf1ad2b        3cb09bc10764                                    "npm start"              41 hours ago        Up 41 hours         3002/tcp                  97c0acf1ad2b_97c0acf1ad2b_web_1
```

**Comment by BenHall** [comment]
Host looks to be completely confused now. Unable to start containers again.

```
Mar 06 14:36:27 www2 dockerd[16648]: time="2017-03-06T14:36:27.265739055+01:00" level=error msg="Handler for DELETE /v1.26/containers/web_1 returned error: No such container: web_1"
Mar 06 14:36:29 www2 dockerd[16648]: time="2017-03-06T14:36:29.514931570+01:00" level=error msg="Handler for POST /v1.21/containers/be487803bd2de234f46643c91bfa86492b01bbbcfcb5e518f1ae97f47a5fa03b/start returned error: service endpoint with name web_1 already exists"
```

**Comment by BenHall** [comment]
If I manually launch it with the same name it still errors. If I rename the container then it's OK.

**Comment by justincormack** [comment]
@BenHall we have a PR for lazy umount https://github.com/docker/docker/pull/31391 that may help here. Something has a file in `shm` open that is reluctant to be killed, according to the original message.

The later ones are odd; I think there may have been some other cleanup fixes cc @cpuguy83

**Comment by cpuguy83** [comment]
The shm stuff is weird. We definitely see it on CI, though.
Interesting this is happening on kernel 4.9, I wonder if this is a new mount bug in 4.9... for that matter I wonder what kernel version we're running in CI.

See https://github.com/docker/docker/pull/31012#issuecomment-284178025

**Comment by cpuguy83** [comment]
@justincormack btw, we already lazy unmount the shm mounts.

**Comment by BenHall** [comment]
Thanks for the updates :)

**Comment by krasi-georgiev** [comment]
once the merge is complete I will test it and troubleshoot if it still causes issues.

**Comment by BenHall** [comment]
Update: 4.9.11 seemed to have completely killed one of our servers. Docker can't even start containers with unique names. The others on older kernels are still happy.

Upgrading to 4.10.1.

**Comment by BenHall** [comment]
Update on this. Docker now has a habit of just refusing to start or stop containers. It completely halts, no debug or logs...

**Comment by BenHall** [comment]
```
$ docker version
Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 11:02:43 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 11:02:43 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

```
$ uname -a
Linux host03 4.10.1-041001-generic #201702260735 SMP Sun Feb 26 12:36:48 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by justincormack** [comment]
I don't recommend just changing kernels all the time, as they have a habit
of adding new bugs as well as fixing old ones. To debug, you need to change
as few things as possible at the same time, you need reproducibility and
you need to use the tools. Device or resource busy is a very different
issue from a hang, you should open a new issue for it. 4.10.1 is not the
current 4.10 kernel and I would only recommend 4.10 if it is solving
problems for you.

On 19 Mar 2017 16:47, "Ben Hall" <notifications@github.com> wrote:

$ docker version
Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 11:02:43 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 11:02:43 2017
 OS/Arch:      linux/amd64
 Experimental: false

$ uname -a
Linux host03 4.10.1-041001-generic #201702260735 SMP Sun Feb 26
12:36:48 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/docker/docker/issues/31195#issuecomment-287629719>, or mute
the thread

**Comment by BenHall** [comment]
I've certainly been trying. I only changed the Kernel on one machine to see if it solved the problem.

As part of debugging, I hit an issue with viewing disk usage via `df -h`. Via strace I found it was blocking when trying to access a /proc directory (`stat("/proc/sys/fs/binfmt_misc"`), I also couldn't ls the files in the directory. 

Restarting proc-sys-fs-binfmt_misc.mount fixed the `df -h` issue. Other github issues have mentioned it could be related to "prometheus/node-exporter".

I have this running on all the hosts as a Docker Container, and could well be when I started seeing an increase in the shm issues. I will continue to investigate.

Sorry for being frustrated in my earlier comment.
